bindings/rust/wasmedge-sys/src/vm.rs

```diff
@@ -0,0 +1,174 @@
+use super::wasmedge;
+use crate::{
+    raw_result::{decode_result, ErrReport},
+    string::StringRef,
+    value::Value,
+    wasi,
+};
+use std::os::raw::c_char;
+
+// If there is a third-party sdk based on wasmedge-sys
+// the private encapsulation  here can force the third-party sdk to use the api
+#[derive(Debug)]
+pub struct Vm {
+    pub(crate) ctx: *mut wasmedge::WasmEdge_VMContext,
+    _private: (),
+}
+
+impl Vm {
+    pub fn create(config: &crate::config::Config) -> Result<Self, ErrReport> {
+        let ctx = unsafe {
+            wasmedge::WasmEdge_VMCreate(config.ctx, std::ptr::null_mut() /* store */)
+        };
+        if ctx.is_null() {
+            panic!("WasmEdge VM create failed");
+        }
+        Ok(Self { ctx, _private: () })
+    }
+
+    pub fn load_wasm_from_ast_module(
+        self,
+        module: &crate::module::Module,
+    ) -> Result<Self, ErrReport> {
+        unsafe {
+            decode_result(wasmedge::WasmEdge_VMLoadWasmFromASTModule(
+                self.ctx, module.ctx,
+            ))?;
+        }
+        Ok(self)
+    }
+
+    pub fn validate(self) -> Result<Self, ErrReport> {
+        unsafe {
+            decode_result(wasmedge::WasmEdge_VMValidate(self.ctx))?;
+        }
+        Ok(self)
+    }
+
+    pub fn instantiate(self) -> Result<Self, ErrReport> {
+        unsafe {
+            decode_result(wasmedge::WasmEdge_VMInstantiate(self.ctx))?;
+        }
+        Ok(self)
+    }
+
+    pub fn init_wasi_obj(
+        &self,
+        args: Option<Vec<&str>>,
+        envs: Option<Vec<&str>>,
+        dirs: Option<Vec<&str>>,
+        preopens: Option<Vec<&str>>,
```

Comment:
why are these all `Option`s, though? An empty `Vec` is as good and gives a cleaner API. `vec![]` doesn't allocate FYI. Also, you might want to do `&[impl AsRef<str>]`  on these in case anyone has a vec of `String` or slice of `str` etc. Going from `Vec<String>` to `Vec<&str>` does require an allocation and is annoying to write. Also, you don't seem to need to take ownership of the args anyway.

Reply:
This is only a temporary solution, and it needs to be refactored later with specific usage scenarios. This should be the goal of version 0.3 .